### ISSUE TITLE
How to add custom plugins?

### ISSUE THREAD

**Issue by CPUFronz** [vanderschaarlab/synthcity#212]
## Question
What is the best way to add custom plugins to `Benchmarks.evaluate()`? 

## Further Information
`Benchmarks.evaluate()` is a really nice interface to run different experiments, but how do I add my own plugins to run experiments with them? In order to get it working, I added the parameter `plugins_context` to `Benchmarks.evaluate()`, but that seemed a little bit much for a rather simple task. Am I missing something?

Here is the snippet:

```python
plugins_context = Plugins()
plugins_context.add('myplugin', MyPlugin)

score = Benchmarks.evaluate(
    [
        ('myplugin', 'myplugin'),
    ],
    loader,
    synthetic_size=1000,
    metrics={
        "performance": ['linear_model', 'xgb', 'detection_xgb'],
    },
    repeats=1,
    plugins_context=plugins_context,
)
Benchmarks.print(score)
```

In `syncity.benchmark.Benchmarks` I added `plugins_context` to the constructor and replaced the two times `Plugins()` is called with `plugins_context`.

## System Information
 - OS: any
 - OS Version: any
 - Language Version: any
 - Package Manager Version: any

**Comment by robsdavis** [comment]
Hi @CPUFronz,

Thanks for engaging with synthcity - I'm glad you are enjoying using the Benchmarks interface.

Yes, there does seem to be an unnecessary limitation here.

For now the easiest ways are to do what you describe above or ...

If you have cloned the repo and are working on it directly, all you need to do is to save you're plugin class to a file here: `synthcity/src/synthcity/plugins/*your_plugin_category*/plugin_*your_plugin_name*`, where `your_plugin_category` should be `"generic"` or one of the other categories (`"privacy"`, `"time_series"` etc.) and `your_plugin_name` can be anything you like. This will allow it to be picked up by the `PluginLoader` and it will become available to `Benchmarks.evaluate()`.

However, I can't see a reason why you should not be able to benchmark custom plugins at runtime. I will consider how to make this easier, perhaps by adding an additional optional argument for plugins_context as you have done above.

**Comment by DrShushen** [comment]
@robsdavis please have a look at temporai's plugin registration code - this will automatically add any dynamically user-registered plugins to be available in benchmarking:
* https://github.com/vanderschaarlab/temporai/blob/main/src/tempor/plugins/core/_plugin.py#L119
* https://github.com/vanderschaarlab/temporai/blob/main/tutorials/extending/tutorial01_custom_plugin.ipynb

**Comment by robsdavis** [comment]
Thanks @DrShushen - that is a very neat way to handle this situation. I will create a specific issue to implement an equivalent solution for synthcity using a plugin registry variable and a `_refresh` method.

**Comment by robsdavis** [comment]
Hi all and @CPUFronz,

This is now implemented in the mentioned PR (#225), such that now calling `plugins().add()` will add your new plugin to the plugin registry, which will make it available to `Benchmarks.evalute()`.

**Comment by CPUFronz** [comment]
Awesome, thanks!
